### PR TITLE
TYP: stronger typing in libindex

### PR DESCRIPTION
**Pull request by jbrockmendel**
and more consistent naming/args for MultiIndex engine.

Preliminary before making index.pyi

### CODE REVIEW

File: pandas/_libs/index.pyx [hunk 1]
```diff
@@ -613,8 +613,8 @@ cdef class BaseMultiIndexCodesEngine:
         lab_ints = self._extract_level_codes(target)
         return self._base.get_indexer(self, lab_ints)
 
-    def get_indexer(self, object target, object values = None,
-                    object method = None, object limit = None) -> np.ndarray:
+    def get_indexer_with_fill(self, ndarray target, ndarray values,
```
Comment (simonjayhawkins) [inline pandas/_libs/index.pyx:616]: docstring update for target

**Review by simonjayhawkins — APPROVED** [review]
Thanks @jbrockmendel lgtm

File: pandas/_libs/index.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/index.pyx:616]: updated

**Comment by simonjayhawkins** [comment]
Thanks @jbrockmendel

**Pull request merged**